File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        index = 0
        # dummy = ListNode(-1)
        # dummy.next = head
        curr = head
        while curr:
            curr = curr.next
            index += 1
        
        ranges = index - (n + 1)
        if ranges  == -1:
            head = head.next
            return head
        print(ranges)
        
        # node = head
        # while node:
        #     print(node.val)
        #     node = node.next
        # print(index)
        
        curr = head
        while ranges > 0:
            curr = curr.next
            ranges -= 1
            
        if not head or not head.next:
            head = None
        elif curr and curr.next:
            curr.next = curr.next.next
            
        return head
            
        

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *temp=new ListNode(0);
        temp->next = head;
        ListNode *slow=temp;
        ListNode *fast=temp;
        
        int x=0;
        while(x<n+1 && fast)
        {
            fast=fast->next;
            x++;
        } 
        while(fast != NULL)
        {
            fast=fast->next;
            slow=slow->next;
        }
        cout<<slow->val<<endl;
        if (slow && slow->next)
        {
            slow->next = slow->next->next;
        }
        
        return temp->next;
        
    }
};
